Resolve and install package dependencies

hasheddan opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Introduces automatic package dependency resolution. A package that is
installed with declared dependencies will have those dependencies
installed on its behalf, as long as they do not conflict with existing
dependencies our cause a cycle.

Please see individual commits for details.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

Fixes #

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

Tested by installing a package with the following dependency tree:

![Dependency Trees](https://user-images.githubusercontent.com/31777345/100272581-05a41e00-2f21-11eb-94b8-00fc4fa211fe.png)


Output:

```
🤖 (crossplane) k get pkg
NAME                                                  INSTALLED   HEALTHY   PACKAGE                                               AGE
provider.pkg.crossplane.io/crossplane-provider-aws    True        True      registry.upbound.io/crossplane/provider-aws:v0.14.0   117s
provider.pkg.crossplane.io/crossplane-provider-helm   True        True      registry.upbound.io/crossplane/provider-helm:v0.3.7   116s

NAME                                                     INSTALLED   HEALTHY   PACKAGE                                             AGE
configuration.pkg.crossplane.io/config-nop-a             True        True      registry.upbound.io/hasheddan/config-nop-a:v0.0.1   2m2s
configuration.pkg.crossplane.io/hasheddan-config-nop-b   True        True      registry.upbound.io/hasheddan/config-nop-b:v0.0.1   2m
configuration.pkg.crossplane.io/hasheddan-config-nop-c   True        True      registry.upbound.io/hasheddan/config-nop-c:v0.0.1   119s
configuration.pkg.crossplane.io/hasheddan-config-nop-d   True        True      registry.upbound.io/hasheddan/config-nop-d:v0.0.1   118s
configuration.pkg.crossplane.io/hasheddan-config-nop-e   True        True      registry.upbound.io/hasheddan/config-nop-e:v0.0.1   117s
```

Followed with installation of `crossplane/provider-gcp`, which does not intersect with dependency tree:

```
🤖 (crossplane) k get pkg
NAME                                                  INSTALLED   HEALTHY   PACKAGE                                               AGE
provider.pkg.crossplane.io/crossplane-provider-aws    True        True      registry.upbound.io/crossplane/provider-aws:v0.14.0   117s
provider.pkg.crossplane.io/crossplane-provider-helm   True        True      registry.upbound.io/crossplane/provider-helm:v0.3.7   116s
provider.pkg.crossplane.io/provider-gcp               True        True      crossplane/provider-gcp:master                        63s

NAME                                                     INSTALLED   HEALTHY   PACKAGE                                             AGE
configuration.pkg.crossplane.io/config-nop-a             True        True      registry.upbound.io/hasheddan/config-nop-a:v0.0.1   2m2s
configuration.pkg.crossplane.io/hasheddan-config-nop-b   True        True      registry.upbound.io/hasheddan/config-nop-b:v0.0.1   2m
configuration.pkg.crossplane.io/hasheddan-config-nop-c   True        True      registry.upbound.io/hasheddan/config-nop-c:v0.0.1   119s
configuration.pkg.crossplane.io/hasheddan-config-nop-d   True        True      registry.upbound.io/hasheddan/config-nop-d:v0.0.1   118s
configuration.pkg.crossplane.io/hasheddan-config-nop-e   True        True      registry.upbound.io/hasheddan/config-nop-e:v0.0.1   117s
```

[contribution process]: https://git.io/fj2m9

hasheddan commented on pkg/controller/pkg/resolver/reconciler.go:
```diff
@@ -0,0 +1,283 @@
+			return reconcile.Result{}, nil
+		}
+
+		// NOTE(hasheddan): we will be unable to fetch tags for private
+		// dependencies because we do not attach any secrets. Consider copying
+		// secrets from parent dependencies.
```
Note that this iteration constrains dependency resolution to packages that can be fetched anonymously. Propagating secrets from a parent node to its dependencies will be addressed in a follow-up.

hasheddan commented on pkg/controller/pkg/revision/dependency.go:
```diff
@@ -0,0 +1,231 @@
+		return total, installed, invalid, errors.Errorf(errMissingDependenciesFmt, missing)
+	}
+
+	// All of our dependencies and transitive dependencies must exist. Check
+	// that neighbors have valid versions.
```
Note that we do not gate on dependencies actually being healthy. As long as they exist we continue with installation. It could be an improvement to check dependency health in the future, but right now it would likely cause unneeded additional reconciles. This could be improved if we are able to stop reconciling every minute for revisions.

negz commented on pkg/controller/pkg/resolver/reconciler.go:
```diff
@@ -0,0 +1,283 @@
+		if err := r.client.Create(ctx, pack); err != nil {
+			log.Debug(errCreateDependency, "error", err)
+			return reconcile.Result{RequeueAfter: shortWait}, nil
+		}
```
Are we relying on the lock matching the actual set of extant packages when we make this create call? Is it possible the two could get out of sync and the create could fail due to the package already existing but not being recorded in the lock?

hasheddan replied:
Yes! That is possible, but that is why we requeue here. If we fail to create, we'll wait a short period then reconcile again, giving the package time to add itself to the `Lock`.

negz commented on pkg/controller/pkg/resolver/reconciler.go:
```diff
@@ -0,0 +1,283 @@
+		// no packagePullSecrets are set. Settings can be modified manually
+		// after dependency creation to address this.
+		pack.SetName(xpkg.ToDNSLabel(ref.Context().RepositoryStr()))
+		pack.SetSource(fmt.Sprintf(packageTagFmt, ref.String(), addVer))
```
Is it worth somehow relating the packages we automatically create to the packages of which they're dependencies? i.e. Adding an owner ref or a label? I'm mostly thinking about making it easy for humans to figure out why a particular package automatically showed up in their system.

hasheddan replied:
Left a note about this a few lines down. I initially was adding a controller ref to the `Lock`, but it didn't feel right that if the `Lock` disappeared that the packages installed would also. I think a label could make sense :+1:

negz reviewed (approved):
This looks great @hasheddan! I left a bunch of comments, but none of them are blockers.

hasheddan commented on apis/pkg/v1alpha1/lock.go:
```diff
@@ -0,0 +1,124 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Packages []LockPackage `json:"packages,omitempty"`
```
I think it would be somewhat of a violation of its purpose if we did. I would prefer that we reflect any information as events on revisions or conditions in their status.

hasheddan commented on pkg/controller/pkg/revision/reconciler_test.go:
```diff
@@ -200,13 +222,53 @@ func TestReconcile(t *testing.T) {
 			},
 		},
+		"ErrDeletedRemoveSelf": {
+			reason: "We should requeue after short wait if revision is deleted and we fail to remove finalizer.",
```
Nope, good catch :)

hasheddan commented on pkg/dag/dag.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+// TraceNode returns a node's neighbors and all transitive neighbors using depth
+// first search.
```
We don't use this to choose which dependency version to install, so the fact that we use DFS is of no real impact to the user.

hasheddan commented on pkg/dag/dag.go:
```diff
@@ -0,0 +1,238 @@
+}
+
+// Init initializes a MapDag and implies missing destination nodes. Any implied
+// nodes are returned. Any existing nodes are cleared.
```
I think I may actually opt for just initializing the `nodes` map in the `NewMapDag()` function as the intention is that someone should be able to add a node without having to call `Init`

hasheddan commented:
@negz Thanks for your review! I believe I have addressed all of your comments by updating or responding. I anticipate needing some follow-on work here so feel free to respond to any of my comments and I'll make sure to address anything in subsequent PRs :+1:
